allow -loadblock blocks to be unsorted

LarryRuane opened the pull request:
This PR was suggested in [this comment](https://github.com/bitcoin/bitcoin/pull/19594#issuecomment-721321744) in #19594, and builds on that PR (the first commit here).

This PR allows the blocks in files specified by the `-loadblock=file` command line (configuration) options (there can be multiple) to be unsorted. This is already allowed by `-reindex`, but currently when `-loadblock` encounters a block with an unknown parent, it is ignored. This means the blocks must be sorted by height to be loaded successfully. This PR fixes this restriction, allowing the blocks across all the specified blocks files to be in any order. This makes the `-loadblock` option more useful. Also, the affected code is cleaner and more maintainable.

(I'll simplify this description move details to a separate comment before this PR gets merged.)

hebasto committed:
refactor: Make mapBlocksUnknownParent local, and rename it

LarryRuane commented:
A useful follow-on to this PR, since loading unordered blocks would become possible, might be the addition of a `-loadblocksdir=` argument that takes a directory path and loads all blocks found in the directory. That way it's not necessary to specify a very large number of `-loadblock` arguments. It would then be easy to import an entire blockchain. 

Another way to accomplish approximately the same thing would be to replace the entire `blocks` directory in the data directory and then `-reindex`. But using `-loadblocksdir` would allow you to "merge" blocks into your existing blocks list (suppressing duplicates), which is kind of cool.

adamjonas commented:
Appears that the CI picked up a couple of argument issues with the `LoadExternalBlockFile` fuzz test:

```
test/fuzz/load_external_block_file.cpp:33:9: error: no matching function for call to 'LoadExternalBlockFile'
        LoadExternalBlockFile(Params(), fuzzed_block_file, &flat_file_pos, &blocks_with_unknown_parent);
        ^~~~~~~~~~~~~~~~~~~~~
./validation.h:162:6: note: candidate function not viable: requires 5 arguments, but 4 were provided
void LoadExternalBlockFile(
     ^
test/fuzz/load_external_block_file.cpp:35:9: error: no matching function for call to 'LoadExternalBlockFile'
        LoadExternalBlockFile(Params(), fuzzed_block_file);
        ^~~~~~~~~~~~~~~~~~~~~
./validation.h:162:6: note: candidate function not viable: requires 5 arguments, but 2 were provided
void LoadExternalBlockFile(
     ^
2 errors generated.
```

LarryRuane commented:
I don't know if anyone has been confused by `-loadblock` not loading all the blocks, but that did happen on the zcash network (a bitcoin core fork that I also work on): https://github.com/zcash/zcash/issues/4679

LarryRuane commented:
Thanks, @adamjonas, force-pushed to fix the fuzz test ([diff](https://github.com/bitcoin/bitcoin/compare/37408b8aab677907256b5fb953bda7ff6ab686f3..7afedcfc716f497601925dd713c89cb835ed5477)).

mjdietzx commented on src/validation.cpp:
```diff
@@ -4670,20 +4681,48 @@ bool LoadGenesisBlock(const CChainParams& chainparams)
+    std::multimap<uint256, FlatFilePos> blocks_with_unknown_parent;
+    for (size_t n_file = 0; n_file < blk_paths.size(); ++n_file) {
+        LogPrintf("%s block file %s ...\n", action, blk_paths[n_file].filename().string());
+        FILE* file = fsbridge::fopen(blk_paths[n_file], "rb");
+        if (file == nullptr) {
+            LogPrintf("%s: Warning: Could not open blocks file %s\n", __func__, blk_paths[n_file]);
+            return;
+        }
```
Isn't this kinda duplication of what `ReadBlockFromDisk` does? I'm wondering why `FlatFilePos pos(n_file, 0);` isn't created, passed to  `ReadBlockFromDisk` (like before, when signature took `FlatFilePos`), and then you can pass it to `LoadExternalBlockFile`

LarryRuane replied:
This duplication exists even without this PR. Before this PR, this `fopen()` happens [here](https://github.com/bitcoin/bitcoin/pull/20331/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dL729). 

To summarize why this happens in two places: There are two ways blocks are "loaded" into memory when reindexing (and, now, with this PR, when using `-loadblock`). First, a blocks file is opened ([here](https://github.com/bitcoin/bitcoin/pull/20331/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R4693)) and read from disk (using large reads) sequentially. At this point, we don't know the individual blocks' starting offsets, so we can't create a `FlatFilePos` to pass to `ReadBlockFromDisk()` as you suggest. Besides, doing these large reads is more efficient, typically many blocks per read -- plus a partial block at the end (since, again, the block boundaries are unknown). Once in memory, the block boundaries are discovered, and each block is deserialized. But the blocks in these blocks files are mostly out of order, so most of the blocks can't be accepted (have `AcceptBlock()` called on it) because its parent has not yet been seen.

When such a parent is later found, the child block is then _re-read_ from disk using `ReadBlockFromDisk()`, deserialized a second time, and then passed to `AcceptBlock()` (allowed because its parent is now present). In this case, we're reading exactly a single block, because we know where it starts (in the file) and its length (which is what a `FlatFilePos` encodes).

mjdietzx commented on src/validation.h:
```diff
@@ -152,12 +152,20 @@ extern const std::vector<std::string> CHECKLEVEL_DOC;
-    FILE* fileIn,
-    FlatFilePos* dbp = nullptr,
-    std::multimap<uint256, FlatFilePos>* blocks_with_unknown_parent = nullptr);
+    const std::vector<fs::path>& blk_paths,
+    size_t n_file,
```
Given the name `LoadExternalBlockFile` I'm confused why `blk_paths` plural is passed as a param, instead of just a single file/path

LarryRuane replied:
That function, `LoadExternalBlockFile()`, does (attempt to) load the blocks from exactly one block file, _but_, doing so may cause, as a side effect, the loading of blocks seen earlier but that could not be loaded then because their parent was missing (and is now present, see my earlier comment). This requires reaching back and reading specific blocks from earlier files, and that requires the full list of block file paths (the `blk_paths` argument). This basic flow is the same as before this PR. I agree this is somewhat confusing, maybe a comment would help. But maybe not, because perhaps there's no shortcut to just studying the code, I'm not sure.

mjdietzx commented on src/validation.cpp:
```diff
@@ -1143,35 +1143,46 @@ static bool WriteBlockToDisk(const CBlock& block, FlatFilePos& pos, const CMessa
-    CAutoFile filein(OpenBlockFile(pos, true), SER_DISK, CLIENT_VERSION);
-    if (filein.IsNull())
-        return error("ReadBlockFromDisk: OpenBlockFile failed for %s", pos.ToString());
+    FILE* file = fsbridge::fopen(path, "rb");
+    if (!file) {
+        return error("ReadBlockFromDisk: open failed for %s", path.string());
+    }
+    if (offset > 0 && fseek(file, offset, SEEK_SET)) {
+        fclose(file);
+        return error("ReadBlockFromDisk: fseek failed for %s offset %u", path.string(), offset);
+    }
+    CAutoFile filein(file, SER_DISK, CLIENT_VERSION);
```
Note: I'm guessing this change was made bc of what you do at line 4810: `ReadBlockFromDisk(*pblockrecursive, blk_paths[child_pos.nFile], child_pos.nPos, chainparams.GetConsensus())`

But I'm failing to understand why you can't just do: `ReadBlockFromDisk(*pblockrecursive, child_pos, chainparams.GetConsensus())`

I'm probably missing something here, but I just want to check

LarryRuane commented on src/validation.cpp:
```diff
@@ -49,6 +49,7 @@
 #include <validationinterface.h>
 #include <warnings.h>
 
+#include <cassert>
```
This isn't needed, I'll remove in a later force-push.

LarryRuane commented on src/test/fuzz/load_external_block_file.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <validation.h>
 
 #include <cstdint>
+#include <map>
```
This isn't needed, I'll remove in a later force-push.

LarryRuane reviewed (commented):
@mjdietzx, thank you for your review!

LarryRuane commented:
Here's a manual test procedure (I'll describe testnet because it's faster, but mainnet works too)
- testnet node is not running
- wait for your testnet3 data directory to get out of date (don't run the client for at least several days)
- change directory to your `.bitcoin` data directory
- rename `testnet3` to `testnet3-save`
- start the node with `-testnet`, wait for it to sync (IBD), it will have more blocks (and more blocks files) than `testnet3-save`
- stop the node
- rename `testnet3` to `testnet3-new`
- rename `testnet3-save` to `testnet3`
- start the node as follows, except replace `191`  with the highest-numbered file in the `testnet3-save` directory:

```
bitcoind -testnet -networkactive=0 $(seq 0 191|xargs printf '-loadblock=testnet3-new/blocks/blk%05u.dat ')
```
This generates a command line like:
```
bitcoind -testnet -networkactive=0 -loadblock=testnet3-new/blocks/blk00000.dat -loadblock=testnet3-new/blocks/blk00001.dat ...
```


- observe that, even though not syncing with peers,  the node successfully reaches a recent block height (look at the most recent `UpdateTip` in `debug.log`).

promag commented:
@LarryRuane I guess it's possible to make a functional test then?

LarryRuane commented:
@promag I didn't write a new functional test, but I did [modify](https://github.com/bitcoin/bitcoin/pull/20331/files#diff-cabb34205d48861dbb53b2ad0df92776bf7d917150caf17778e15fbc8e63e123R73) the existing loadblock functional test to cover the new behavior (the test fails without the PR). Is that sufficient?

dongcarl commented:
Verified that the test fails without the new behaviour, cool!

A couple of things that could be done:
1. Instead of asserting based on the debug log (which is a bit more fragile), perhaps the assertion can be made based on the number of blocks connected (tip height). We may need to disable networking for this to work reliably.
2. I think the changes in `test/functional` should be split into its own commit and put as the first commit in this patchset, this way it's easier for reviewers to verify that the new behavior does indeed do what it's supposed to do.

LarryRuane committed:
refactor: Add a new version of `ReadBlockFromDisk()`

This lower-level version takes a path and a seek offset as arguments.
Currently unused but is needed by the next commit.

LarryRuane commented:
Force-pushed to implement both suggestions from @dongcarl (thank you!).

Reminder, this PR is layered on top of #19594 so please review only the last _two_ commits (of the 3).

LarryRuane commented:
Rebased, no functional changes.

LarryRuane committed:
allow -loadblock blocks to be unsorted

Add LoadExternalBlockFiles() (plural) that accepts a list of paths and
loads all the blocks found in these paths. The blocks can be completely
out of order; block C (child) can be seen first and its parent P not
seen until a later file, then block C will be processed.  This requires
re-reading block C from the earlier file; block C isn't kept in memory
during this interval.

The list of paths is either the blocks/blk*.dat files (during reindexing)
or given with one or more -loadblk arguments.

LoadExternalBlockFiles() calls LoadExternalBlockFile() (singular) to
read a single file, but out-of-order blocks requires that this function
be able to read individual blocks from earlier files.

LarryRuane commented:
This required rebase also splits the second commit into two commits to separate refactoring from the functional changes. Ready for review.

LarryRuane commented:
Rebased. The second commit is now much simpler because of #21727.

sipa commented:
utACK dbd7f729f42d58a1d47700dc75fb1a5a6d76cdf2

LarryRuane commented:
Rebased to resolve merge conflicts.

LarryRuane commented:
The `-loadblock` feature doesn't seem to be often used, so improvements to it aren't much in demand, so closing this. If this feature becomes more important, we can resurrect this PR.
